### DISCUSSION TITLE
Shelly Pro 1PM

### DISCUSSION THREAD

**Discussion by nduhamel** [Lukeskaiwalker/Shelly-Pro-1-ESPHome#2] [General]
Hi,

Thank for your work on the Shelly Pro 1.
I have a Shelly Pro 1PM and i want to flash it with ESPHome.
It's the same chip as the Pro 1, I post pictures and details on this repo: https://github.com/nduhamel/shelly-pro-1PM-esphome.
I hope you could help me to figure out how you find the correct GPIO and how the ADE7953 power meter is connected on my board.
I have to admit that I don't know where to start.

Thank you for your help.
Nicolas

**Comment by Lukeskaiwalker** [comment]
I have also gotten my hands on one of them and can check it out some time soon. Sadly one inductor broke off from the psu unit and I need to replace it so that I can do some further device testing. 
But I think they also changed the pcb layout slightly. 
Maybe next week I can send you some more infos about a possible pinout. 

Have a nice rest of your weekend. 
![image](https://user-images.githubusercontent.com/38843794/184502757-20b40193-f023-45aa-966c-7eb6472ca9c1.jpeg)

**Comment by nduhamel** [comment]
Thank you for your reply. I will keep you update if i make progress.

Have a nice weekend too.

**Comment by Lukeskaiwalker** [comment]
So I've actually done some reverse engineering and found a probable pinout which I have not tester yet. But if you got some time you could check out if it works

ADE7935 | ESP 32
--------|---------
SDA     | GPIO 12
SCL     | GPIO 13
CS      | GPIO 2
SCLK    | GPIO 14


You might notice that some of those pins also get used by the shift register, but there are connected to each other, so you can try it with both components using the same pin and maybe it works. Also ESP Home only supports the power measuring ic via I2C so the CS and SCLK pin should be irrelevant. The IRQ pin I have not figured out at this point but it looks like it is connected to somewhere. 

Hope that it might work. If it does or even if it doesn't you can fill me in with what you've found out. 

Greetings and have a nice day.

**Reply by Lukeskaiwalker** [reply]
Update: 
Found the IRQ pin 
It is connected to the ESPs GPIO 16 Pin.

**Comment by nduhamel** [comment]
Hi,

I just tried your pinout.
The bad news, i can't read data from the ADE7953, i also try GPIO14 as SCL (same clock as the sn74hc595).
Good news, everything else work as expected or at least the relay, the switch and the rgb led.

Here the yaml, i used:
```
esphome:
  name: shelly-pro-1-pm

esp32:
  board: esp32dev
  framework:
    type: arduino

# Enable logging
logger:
  level: DEBUG

debug:
  update_interval: 5s

text_sensor:
  - platform: debug
    device:
      name: "Device Info"

packages:
  wifi: !include common/wifi.yaml
  mqtt: !include common/mqtt.yaml

web_server:
  port: 80
  include_internal: true

ota:
  safe_mode: true

light:
  - platform: status_led
    name: "Switch state"
    pin:
      sn74hc595: sn74hc595_hub
      # Use pin number 1
      number: 1
      inverted: true

#ethernet:
#  type: LAN8720
#  mdc_pin: GPIO23
#  mdio_pin: GPIO18
#  clk_mode: GPIO17_OUT

i2c:
  sda: GPIO12
  # scl: GPIO13
  scl: GPIO14

sensor:
  - platform: ade7953
    irq_pin: GPIO16
    voltage:
      name: Shelly Voltage
    current_a:
      name: Shelly Current B
    current_b:
      name: Shelly Current A
    active_power_a:
      name: Shelly Active Power B
      filters:
        - multiply: -1
    active_power_b:
      name: Shelly Active Power A
      filters:
        - multiply: -1
    update_interval: 3s

button:
  - platform: restart
    id: restart_1

binary_sensor:

  - platform: status
    id: status_1
    on_press:
      then:
        switch.turn_on: rgb_b
    on_release:
      then:
        switch.turn_off: rgb_b

  - platform: gpio
    id: reset_button
    pin:
      number: 35
      inverted: true
    on_click:
      min_length: 400ms
      max_length: 800ms
      then:
        button.press: restart_1
    on_double_click:
      then:
        switch.toggle: light1

  - platform: gpio
    id: input1
    pin:
      number: 38
    on_press:
      then:
        switch.toggle: light1

  - platform: gpio
    id: input2
    name: "Shelly Pro Button 2"
    pin:
      number: 39

sn74hc595:
  - id: 'sn74hc595_hub'
    data_pin: GPIO13
    clock_pin: GPIO14
    latch_pin: GPIO4
    sr_count: 1

switch:
  - platform: gpio
    name: "Shelly Pro Relay"
    id: light1
    pin:
      sn74hc595: sn74hc595_hub
      # Use pin number 0
      number: 0
      inverted: false

  - platform: gpio
    id: out2
    pin:
      sn74hc595: sn74hc595_hub
      # Use pin number 1
      number: 1
      inverted: false

  - platform: gpio
    id: rgb_b
    pin:
      sn74hc595: sn74hc595_hub
      # Use pin number 2
      number: 2
      inverted: false

  - platform: gpio
    id: rgb_g
    pin:
      sn74hc595: sn74hc595_hub
      # Use pin number 3
      number: 3
      inverted: false

  - platform: gpio
    id: rgb_r
    pin:
      sn74hc595: sn74hc595_hub
      # Use pin number 4
      number: 4
      inverted: false
```

How did you find your pinout ? you trace the pin from the power meter to the esp chip ?

Thank you, 
Have a nice evening

**Comment by Lukeskaiwalker** [comment]
Ok that's what I've expected. I think they normally use the SPI bus of the Chips and then they are able to share the same pin.

I do not have any other solution for you then to try it out with a different software or to write your own component in esp home and then hope that it will work.

I have two ways of doing the tracing. First i take two picture, one of the front and one of the back. Then i overlay them and invert them so that the sides are aligned. I use apple keynote to change the opacity of either image.
![image](https://user-images.githubusercontent.com/38843794/184942494-1dd91255-b16f-4f26-b2bc-1b01e37118e9.jpeg)

The other way to confirm my findings is to use a multimeter with a real small probe to test continuity between the contacts I've found out.
This works pretty good for me.

Also i use pictures of the chips data sheet to mark where each pin is connected to, wich looks like this (a bit messy but good to keep track of everything)
![image](https://user-images.githubusercontent.com/38843794/184943113-7ee95ef7-c97b-45ee-aee8-868ce0f5150a.jpeg)


Hope this was at least a bit helpful.

Greetings.

**Comment by nduhamel** [comment]
> Hope this was at least a bit helpful.

I learned a lot, thanks to you.

> Ok that's what I've expected. I think they normally use the SPI bus of the Chips and then they are able to share the same pin.

There something, i don't understand. The 74HC595 don't use SPI bus, it's just share the clock pin (GPIO 14) ?
And if i don't make mistake for using the SPI bus i need this pinout:

| ADE7935     |    ESP32 |
|------------------|--------------|
| SDA/MISO    |     12 |
| SCL/MOSI     |    13 |
| CS               | 2 |
| SCLK           |  14 |
| IRQ Pin        |  16 |

What the purpose of IRQ pin ?

I found this code https://github.com/CalPlug/ADE7953-Wattmeter/tree/master/library/ESP32_SPI/ADE7953ESP32
I will try to adapt it to ESPHome. I keep you update. 

Have a nice day
